{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f5858f0d-f209-470d-99f1-8d45229aa9d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages({\n",
    "    library(DSS)\n",
    "    library(bsseq)\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    library(BiocParallel)\n",
    "    library(ggplot2)\n",
    "    library(gridExtra)\n",
    "    library(scales)\n",
    "    library(viridis)\n",
    "    library(rtracklayer)\n",
    "    library(GenomicRanges)\n",
    "    library(IRanges)\n",
    "    library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "605d6bb7-b659-49bb-bc6d-a4d16f43dac4",
   "metadata": {},
   "outputs": [],
   "source": [
    "source(\"../scripts/analysis_Util.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e712876d-d54d-4e9b-8007-f706b1865722",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading methylation data files...\n",
      "BSseq object created:\n",
      "An object of type 'BSseq' with\n",
      "  612955 methylation loci\n",
      "  2 samples\n",
      "has not been smoothed\n",
      "All assays are in-memory\n",
      "\n",
      "Running DML test with smoothing...\n",
      "Smoothing ...\n",
      "Estimating dispersion for each CpG site, this will take a while ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in mclapply(1:nrow(X2), foo, mc.cores = ncores):\n",
      "“scheduled core 3 did not deliver a result, all values of the job will be affected”\n",
      "Warning message in shrk.phi[ix] <- shrk.phi2:\n",
      "“number of items to replace is not a multiple of replacement length”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing test statistics ...\n",
      "Smoothed DML test completed. Total tests performed: 605192 \n",
      "\n",
      "Running DML test without smoothing...\n",
      "Estimating dispersion for each CpG site, this will take a while ...\n",
      "Unsmoothed DML test completed. Total tests performed: 588877 \n",
      "\n",
      "P-value summaries:\n",
      "Smoothed analysis:\n",
      "     Min.   1st Qu.    Median      Mean   3rd Qu.      Max. \n",
      "0.0000000 0.0003247 0.0441408 0.2086218 0.3511524 0.9999968 \n",
      "\n",
      "Unsmoothed analysis:\n",
      "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
      " 0.0000  0.1086  0.3709  0.4082  0.6748  1.0000 \n",
      "\n",
      "Calling significant DMLs...\n",
      "\n",
      "--- DML Results Comparison ---\n",
      "Smoothed analysis - Significant DMLs: 112386 \n",
      "Unsmoothed analysis - Significant DMLs: 68783 \n",
      "Criteria: p-value < 0.05 and |methylation difference| > 0.1 \n",
      "\n",
      " Overlapping significant DMLs between methods: 44399 \n",
      "Overlap as % of smoothed DMLs: 39.5 %\n",
      "Overlap as % of unsmoothed DMLs: 64.5 %\n",
      "\n",
      "...Identifying CpG sites lost during smoothing...\n",
      "Common sites between smoothed and unsmoothed: 588877 \n",
      "Percentage of smoothed sites: 97.3 %\n",
      "Percentage of unsmoothed sites: 100 %\n",
      "Length of smoothed indices: 588877 \n",
      "Length of unsmoothed indices: 588877 \n",
      "\n",
      "Found 18423 CpG sites with significant signal loss after smoothing\n",
      "Signal loss sites saved to: HG002_pb_vs_st001.lung_pb_signal_loss_sites.bed \n",
      "\n",
      "Found 3836 sites smoothed to effectively zero\n",
      "Of these, 3836 had only 1-5 methylated reads\n",
      "Sites smoothed to zero saved to: HG002_pb_vs_st001.lung_pb_sites_smoothed_to_zero.bed \n",
      "Low count sites saved to: HG002_pb_vs_st001.lung_pb_low_count_to_zero.bed \n",
      "\n",
      "--- DMLs Lost in Smoothing Analysis ---\n",
      "DMLs significant only in unsmoothed analysis: 24384 \n",
      "Percentage of unsmoothed DMLs lost: 35.5 %\n",
      "Lost DMLs saved to: HG002_pb_vs_st001.lung_pb_DMLs_lost_in_smoothing.bed \n",
      "\n",
      "Top 10 Smoothed DMLs:\n",
      "      chr      pos       mu1       mu2      diff    diff.se     stat       phi1\n",
      "659 chr22 10570890 0.9932558 0.3684853 0.6247705 0.06076589 10.28160 0.05095829\n",
      "676 chr22 10571612 0.7756687 0.4728312 0.3028376 0.02348483 12.89503 0.04978707\n",
      "679 chr22 10571630 0.7754509 0.4755350 0.2999159 0.02150663 13.94528 0.04978707\n",
      "687 chr22 10571687 0.7747313 0.4750340 0.2996974 0.02376543 12.61064 0.04978707\n",
      "688 chr22 10571691 0.7747313 0.4750340 0.2996974 0.02290176 13.08622 0.04978707\n",
      "693 chr22 10571726 0.7524587 0.4771353 0.2753235 0.01723279 15.97672 0.04978707\n",
      "699 chr22 10571763 0.7522486 0.4813911 0.2708576 0.01992142 13.59630 0.04978707\n",
      "700 chr22 10571769 0.7521440 0.4816683 0.2704757 0.01746688 15.48506 0.04978707\n",
      "701 chr22 10571771 0.7520918 0.4802072 0.2718846 0.01823539 14.90972 0.04978707\n",
      "702 chr22 10571780 0.7521681 0.4829904 0.2691776 0.01981470 13.58474 0.04978707\n",
      "          phi2         pval          fdr postprob.overThreshold\n",
      "659 0.04480783 8.529993e-25 7.208886e-23                      1\n",
      "676 0.05240865 4.800717e-38 8.887597e-36                      1\n",
      "679 0.07642668 3.361335e-44 8.507959e-42                      1\n",
      "687 0.03684867 1.844805e-36 3.150286e-34                      1\n",
      "688 0.04530480 3.947852e-39 7.707124e-37                      1\n",
      "693 0.19489049 1.856516e-57 8.435052e-55                      1\n",
      "699 0.03483320 4.212100e-42 9.590404e-40                      1\n",
      "700 0.05080607 4.376252e-54 1.722024e-51                      1\n",
      "701 0.09073760 2.849386e-50 9.543029e-48                      1\n",
      "702 0.04318192 4.932630e-42 1.117629e-39                      1\n",
      "\n",
      "Top 10 Unsmoothed DMLs:\n",
      "        chr      pos       mu1        mu2      diff    diff.se      stat\n",
      "5929  chr22 11265291 0.9568477 0.12619179 0.8306559 0.07194841 11.545160\n",
      "5993  chr22 11267987 0.9617103 0.13987160 0.8218387 0.07072406 11.620356\n",
      "6185  chr22 11273231 0.9512898 0.16838679 0.7829030 0.06792914 11.525289\n",
      "7285  chr22 11353533 0.9278425 0.01682330 0.9110193 0.04677571 19.476332\n",
      "7286  chr22 11353594 0.8989516 0.09818624 0.8007654 0.06637774 12.063765\n",
      "8093  chr22 11489094 0.9845533 0.05549691 0.9290564 0.06491298 14.312336\n",
      "9205  chr22 11615113 0.9008698 0.08353094 0.8173389 0.07610824 10.739164\n",
      "9931  chr22 11698573 0.8967574 0.07111728 0.8256401 0.08261635  9.993664\n",
      "14519 chr22 12196395 0.8791361 0.08343239 0.7957037 0.06755294 11.778966\n",
      "16001 chr22 12423622 0.8909806 0.04113080 0.8498498 0.06622263 12.833224\n",
      "            phi1       phi2         pval          fdr postprob.overThreshold\n",
      "5929  0.01565145 0.01565145 7.809581e-31 2.090401e-26                      1\n",
      "5993  0.01507715 0.01507715 3.247789e-31 9.107372e-27                      1\n",
      "6185  0.01226738 0.01226738 9.838147e-31 2.518895e-26                      1\n",
      "7285  0.01279596 0.01279596 1.743320e-84 1.026601e-78                      1\n",
      "7286  0.01189997 0.01189997 1.641066e-33 6.039913e-29                      1\n",
      "8093  0.02947348 0.02947348 1.832534e-46 5.395686e-41                      1\n",
      "9205  0.01550437 0.01550437 6.664532e-27 9.811474e-23                      1\n",
      "9931  0.01844505 0.01844505 1.624624e-23 1.138933e-19                      1\n",
      "14519 0.01209209 0.01209209 5.010403e-32 1.475255e-27                      1\n",
      "16001 0.01496212 0.01496212 1.068251e-37 1.048447e-32                      1\n",
      "\n",
      "Smoothed DMLs saved to: HG002_pb_vs_st001.lung_pb_smoothed_DMLs.csv \n",
      "Unsmoothed DMLs saved to: HG002_pb_vs_st001.lung_pb_unsmoothed_DMLs.csv \n",
      "\n",
      "Calling DMRs...\n",
      "\n",
      "Top 10 Smoothed DMRs:\n",
      "        chr    start      end length nCG meanMethy1 meanMethy2 diff.Methy\n",
      "1051  chr22 16323773 16332969   9197 466  0.8127383 0.47800826  0.3347301\n",
      "14574 chr22 48748477 48761038  12562 379  0.3609734 0.78565616 -0.4246827\n",
      "5854  chr22 27797284 27800384   3101 274  0.8367684 0.48286890  0.3538995\n",
      "198   chr22 11252224 11255497   3274 314  0.8739585 0.58705770  0.2869008\n",
      "13416 chr22 46262296 46263617   1322 151  0.6755158 0.16141908  0.5140967\n",
      "6516  chr22 29480274 29481424   1151 151  0.5024156 0.07202195  0.4303937\n",
      "1685  chr22 17868095 17884143  16049 321  0.4933667 0.87507611 -0.3817094\n",
      "194   chr22 11248798 11250981   2184 211  0.8177293 0.48457982  0.3331495\n",
      "341   chr22 11627454 11629744   2291 222  0.9467510 0.74879045  0.1979606\n",
      "15431 chr22 50546387 50547824   1438 110  0.8869230 0.37362692  0.5132961\n",
      "       areaStat\n",
      "1051   4332.046\n",
      "14574 -3883.243\n",
      "5854   3527.621\n",
      "198    3377.884\n",
      "13416  3147.606\n",
      "6516   2836.931\n",
      "1685  -2672.133\n",
      "194    2263.688\n",
      "341    2166.460\n",
      "15431  2150.790\n",
      "\n",
      "Top 10 Unsmoothed DMRs:\n",
      "        chr    start      end length nCG meanMethy1 meanMethy2 diff.Methy\n",
      "31586 chr22 46262841 46263427    587  88 0.72329247 0.13824598  0.5850465\n",
      "9989  chr22 23798974 23801596   2623  62 0.14656347 0.83047895 -0.6839155\n",
      "37202 chr22 50546741 50547625    885  87 0.88307755 0.32066019  0.5624174\n",
      "35228 chr22 49051776 49052733    958  56 0.12906071 0.77425057 -0.6451899\n",
      "15225 chr22 29480855 29481175    321  50 0.76813576 0.09007738  0.6780584\n",
      "24501 chr22 39441810 39443458   1649  40 0.09281914 0.86188401 -0.7690649\n",
      "4256  chr22 17800108 17801448   1341  44 0.07370389 0.80368756 -0.7299837\n",
      "13897 chr22 27798053 27798853    801  83 0.92263902 0.54943577  0.3732033\n",
      "32508 chr22 47090577 47091602   1026  49 0.80740466 0.23046057  0.5769441\n",
      "32262 chr22 46768313 46769553   1241  38 0.10850800 0.76022789 -0.6517199\n",
      "       areaStat\n",
      "31586  434.1964\n",
      "9989  -362.9985\n",
      "37202  354.1790\n",
      "35228 -327.7596\n",
      "15225  293.9441\n",
      "24501 -281.3418\n",
      "4256  -249.4929\n",
      "13897  233.4042\n",
      "32508  232.0925\n",
      "32262 -228.5771\n",
      "\n",
      "Smoothed DMRs saved to: HG002_pb_vs_st001.lung_pb_smoothed_DMRs.csv \n",
      "Unsmoothed DMRs saved to: HG002_pb_vs_st001.lung_pb_unsmoothed_DMRs.csv \n",
      "\n",
      "--- DMR Comparison ---\n",
      "Method comparison:\n",
      "Smoothed DMRs:  8508 \n",
      "Unsmoothed DMRs:  3422 \n",
      "Identical DMRs between methods: 117 \n"
     ]
    }
   ],
   "source": [
    "methyOut <- run_methSmooth_analysis(\"../data/HG002.pacbio.chr22.count.dss.bed\", \"../data/st001.lung.pacbio.chr22.count.dss.bed\",\n",
    "                                    \"HG002_pb\", \"st001.lung_pb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2865418a-91a8-4c7f-8c57-82e335b6803b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating combined axis visualizations...\n",
      "Creating overlaid DML p-value distribution...\n",
      "Creating overlaid DML methylation difference distribution...\n",
      "Methylation difference distribution plots saved to:\n",
      " ./plots/HG002_pb_vs_st001.lung_pb_DML_methylation_diff_overlaid_wide.png \n",
      "Creating overlaid volcano plot...\n",
      "Volcano plot saved to:\n",
      " ./plots/HG002_pb_vs_st001.lung_pb_volcano_plot_overlaid_wide.png \n",
      "Creating DML count per chromosome bar chart...\n",
      "DML per chromosome plots saved to:\n",
      " ./plots/HG002_pb_vs_st001.lung_pb_DML_count_per_chromosome_wide.png \n",
      "Creating overlaid DMR length distribution...\n",
      "DMR length distribution plots saved to:\n",
      " ./plots/HG002_pb_vs_st001.lung_pb_DMR_length_overlaid_wide.png \n",
      "Creating overlaid DMR methylation difference distribution...\n",
      "DMR methylation difference distribution plots saved to:\n",
      " ./plots/HG002_pb_vs_st001.lung_pb_DMR_methylation_diff_overlaid_wide.png \n",
      "Creating visualization of DMLs lost in smoothing...\n",
      "DMLs lost in smoothing plot saved to: ./plots/HG002_pb_vs_st001.lung_pb_DMLs_lost_in_smoothing_wide.png \n",
      "Creating combined grid of all plots...\n",
      "Combined plot grid saved to: ./plots/HG002_pb_vs_st001.lung_pb_methylation_analysis_grid.png \n"
     ]
    }
   ],
   "source": [
    "plotss <- create_plots(methyOut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0c273ef6-069a-4b62-b9a3-2ad00d8539fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using genome build: hg38 \n"
     ]
    }
   ],
   "source": [
    "source(\"..//scripts//createBigWigs.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c88d0d8c-4565-4c9f-a7d6-ff7915f068ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== Creating Methylation Difference Tracks ===\n",
      "Created: HG002_pb_vs_st001.lung_pb_methylation_difference_smoothed.bw \n",
      "Created: HG002_pb_vs_st001.lung_pb_methylation_difference_unsmoothed.bw \n",
      "\n",
      "=== Creating Statistical Significance Tracks ===\n",
      "Created: HG002_pb_vs_st001.lung_pb_significance_smoothed.bw \n",
      "Created: HG002_pb_vs_st001.lung_pb_significance_unsmoothed.bw \n",
      "\n",
      "=== Creating Significant DML Tracks ===\n",
      "Created: HG002_pb_vs_st001.lung_pb_significant_DMLs_smoothed_binary.bw \n",
      "Created: HG002_pb_vs_st001.lung_pb_significant_DMLs_smoothed_weighted.bw \n",
      "Created: HG002_pb_vs_st001.lung_pb_significant_DMLs_unsmoothed_binary.bw \n",
      "Created: HG002_pb_vs_st001.lung_pb_significant_DMLs_unsmoothed_weighted.bw \n",
      "Created: HG002_pb_vs_st001.lung_pb_DMRs_smoothed_methylation_difference.bw \n",
      "Processing smoothed DMR significance track using areaStat...\n",
      "Created: HG002_pb_vs_st001.lung_pb_DMRs_smoothed_significance.bw  (using areaStat)\n",
      "Created: HG002_pb_vs_st001.lung_pb_DMRs_smoothed_CpG_count.bw \n",
      "Created: HG002_pb_vs_st001.lung_pb_DMRs_unsmoothed_methylation_difference.bw \n",
      "Processing unsmoothed DMR significance track using areaStat...\n",
      "Created: HG002_pb_vs_st001.lung_pb_DMRs_unsmoothed_significance.bw  (using areaStat)\n",
      "Created: HG002_pb_vs_st001.lung_pb_DMRs_unsmoothed_CpG_count.bw \n",
      "\n",
      "=== Creating Comparison Tracks ===\n",
      "Created: HG002_pb_vs_st001.lung_pb_DMLs_consensus_both_methods.bw \n",
      "Created: HG002_pb_vs_st001.lung_pb_DMLs_consensus_weighted_average.bw \n",
      "Created: HG002_pb_vs_st001.lung_pb_DMLs_smoothed_specific.bw \n",
      "Created: HG002_pb_vs_st001.lung_pb_DMLs_unsmoothed_specific.bw \n"
     ]
    }
   ],
   "source": [
    "bw <- create_bigwig(methyOut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cb7214f-f630-49b4-a85f-fe12c5e51589",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
